### DISCUSSION TITLE
Run migration(skip if applied) when start avram is connecting.

### DISCUSSION THREAD

**Discussion by zw963** [luckyframework/lucky#1710] [Q&A] [answered]
Hello, i am working on a kemal + avram project which run on raspberry PI (ARM), i have to cross build the intermedia object file from my X86_64 laptop, and link it on PI to create the binary.

this project simple enough anyway, but the issue is, i expect to run migration(if still not) automatically when i start my kemal web server which will load my db config before start.

following is my db config:

```crystal
require "pg"
require "avram"
require "../models/base_model"
require "../models/**"
require "../queries/**"
require "../operations/**"

class AppDatabase < Avram::Database
end

AppDatabase.configure do |settings|
  settings.credentials =
    Avram::Credentials.parse?(ENV["DATABASE_URL"]?) ||
    Avram::Credentials.new(
      database: "#{app_name}_#{app_env}",
      hostname: postgres_host,
      port: postgres_port,
      username: postgres_user,
      password: postgres_password
    )
end

Avram.configure do |settings|
  settings.database_to_migrate = AppDatabase

  settings.lazy_load_enabled = app_env != "production"
end
```

I have two question here:

1. How to run migration on my X86_64 development laptop?
   
   Because i am avram only, this is not a valid lucky project.

   so, when i run `lukcy db.create`, i get following error message
   

```crystal
lucky db.create

You are not in a Lucky project

Try this...

  ▸ Run lucky init to create a Lucky project.
  ▸ Change your project's root directory to see what tasks are available.

```

i can use `bin/lucky.gen.migration` create new migration in `/db/migraitons`,  but i don't know how to run it.

2. How to run those migration on Raspbian OS when do deployment? there is a limit resources, and no crystal compiler available.

Thank you.

**Comment by jwoertink** [comment, marked as answer]
You can either make lucky_tasks a required dependency, then add a tasks.cr in your app root like a Lucky project would do, or you can run the migrations from code with

Avram::Migrator::Runner.new.run_pending_migrations

That will always try to run migrations. You may need to require your migration directory as well.

**Reply by zw963** [reply]
Hi, @jwoertink 

It works after add a new tasks file like this with some empty folders.

```crystal
# tasks.cr

# See `LuckyEnv#task?`
ENV["LUCKY_TASK"] = "true"

require "./src/config/database"
require "lucky_task"
require "./tasks/**"
require "./db/migrations/**"

LuckyTask::Runner.run
```

Add following code to my src/config/database.cr works as well.

```crystal
require "../../db/migrations/**"
Avram::Migrator::Runner.new.run_pending_migrations
```

For my cases, the latter is better, because it run migration automatically when start server.

Really thank you very much for you help.

**Reply by zw963** [reply]
Hi, @jwoertink ,  after i add the `tasks.cr` as above, no `lucky gen.model` task created.

I have to copy follow files from lucky shards into my project to make it works. (with a extra development dependency: teeplate)

```git
A  tasks/db/gen/mixins/migration_with_columns.cr
A  tasks/db/gen/model.cr
A  tasks/db/gen/templates/model/models/{{underscored_namespace_path}}{{underscored_name}}.cr.ecr
A  tasks/db/gen/templates/model/operations/{{underscored_namespace_path}}delete_{{underscored_name}}.cr.ecr
A  tasks/db/gen/templates/model/operations/{{underscored_namespace_path}}save_{{underscored_name}}.cr.ecr
A  tasks/db/gen/templates/model/queries/{{underscored_namespace_path}}{{underscored_name}}_query.cr.ecr
A  tasks/db/gen/templates/model_template.cr
```

Should we move those files into `avram` gem instead of `lucky` ? because really those file i can create manually,
and work well only with `avram`, even no lucky.

Thank you.

**Reply by jwoertink** [reply]
Those are already moved in to Avram and will be there in the next release.

**Reply by zw963** [reply]
> Those are already moved in to Avram and will be there in the next release.

Yes, it works, after switch to main branch, and require ``require "avram/src/lucky/tasks/gen/model"`` manually in tasks.cr, it works.

But, still have two question:

1.  Why add `lib/avram/src/lucky/tasks/gen/route_inferrer.cr` there? is that file relative to avram?
2.  Should we add `db.seed.required_data` and `db.seed.sample_data` into `avram` instead?

Thank you.

**Reply by jwoertink** [reply]
1. I think the route inferer is there to help generate different Avram operations based on the type of route. I don't remember exactly.

2. The db.seed tasks belong in each individual Lucky app since they are specific to your app.

**Reply by zw963** [reply]
> I think the route inferer is there to help generate different Avram operations based on the type of route. I don't remember exactly.

Okay, thank you.

>  The db.seed tasks belong in each individual Lucky app since they are specific to your app.

I think we can create a empty task, which make `lucky db.seed.???` tasks name available, anyway, differentiate those two task can be consider as a best practice, what you think?

**Reply by jwoertink** [reply]
> I think we can create a empty task

We already create empty tasks in the Lucky projects https://github.com/luckyframework/lucky_cli/tree/main/src/web_app_skeleton/tasks/db/seed

We can't put them in Avram though because there's no way to override tasks by name, or guarantee loading order. The tasks have to be generated in the project itself so each dev can update it manually. So for a Kemal + Avram project, this would be something you would do by hand.
